src/session/content/message_row/mod.rs

```diff
@@ -2,149 +2,307 @@ mod sticker;
-        } else {
-            self.set_halign(gtk::Align::Start);
-            self.set_margin_start(0);
-            self.set_margin_end(38);
+    fn set_message(&self, message: Option<glib::Object>) {
+        if self.message() == message {
+            return;
         }
 
-        // Show avatar, if needed
-        let show_avatar = {
-            if !message.is_outgoing() {
-                match message.chat().type_() {
-                    ChatType::BasicGroup(_) => true,
-                    ChatType::Supergroup(data) => !data.is_channel,
-                    _ => false,
+        let self_ = imp::MessageRow::from_instance(self.upcast_ref());
+        if let Some(ref message) = message {
+            if let Some(message) = message.downcast_ref::<Message>() {
+                self_.is_outgoing.set(message.is_outgoing());
+
+                let show_avatar = {
+                    if !message.is_outgoing() {
+                        match message.chat().type_() {
+                            ChatType::BasicGroup(_) => true,
+                            ChatType::Supergroup(data) => !data.is_channel,
+                            _ => false,
+                        }
+                    } else {
+                        false
+                    }
+                };
+                if show_avatar {
+                    let avatar_item = match message.sender() {
+                        MessageSender::User(user) => user.avatar().clone(),
+                        MessageSender::Chat(chat) => chat.avatar().clone(),
+                    };
+
+                    if self_.avatar.borrow().is_none() {
+                        let avatar = Avatar::new();
+                        avatar.set_size(AVATAR_SIZE);
+                        avatar.set_item(Some(avatar_item));
+                        avatar.set_parent(self.upcast_ref());
+                        self_.avatar.replace(Some(avatar));
+                    } else if let Some(avatar) = self_.avatar.borrow().as_ref() {
+                        avatar.set_item(Some(avatar_item));
+                    }
+                } else {
+                    if let Some(avatar) = self_.avatar.borrow().as_ref() {
+                        avatar.unparent();
+                    }
+                    self_.avatar.replace(None);
                 }
+            } else if message.downcast_ref::<SponsoredMessage>().is_some() {
+                self_.is_outgoing.set(false);
             } else {
-                false
+                unreachable!("Unexpected message type: {:?}", message);
             }
-        };
-        if show_avatar {
-            let avatar = if let Some(Ok(avatar)) =
-                self_.avatar_bin.child().map(|w| w.downcast::<Avatar>())
-            {
-                avatar
+        }
+
+        if let Some(content) = self_.content.borrow().as_ref() {
+            if self_.is_outgoing.get() {
+                content.set_margin_start(AVATAR_SIZE + SPACING);
```

Comment:
Margin set, but never reset in the else block. This means that the margin is still present while recycling the widget and the new message is not outgoing. Same thing for the margin set below.